clientv3/retry_interceptor.go

```diff
@@ -104,32 +116,149 @@ func (c *Client) unaryClientInterceptor(logger *zap.Logger, optFuncs ...retryOpt
 		grpcOpts, retryOpts := filterCallOptions(opts)
 		callOpts := reuseOrNewWithCallOptions(intOpts, retryOpts)
-		// short circuit for simplicity, and avoiding allocations.
-		if callOpts.max == 0 {
-			return streamer(ctx, desc, cc, method, grpcOpts...)
-		}
-		if desc.ClientStreams {
-			return nil, status.Errorf(codes.Unimplemented, "clientv3/retry_interceptor: cannot retry on ClientStreams, set Disable()")
-		}
-		newStreamer, err := streamer(ctx, desc, cc, method, grpcOpts...)
-		if err != nil {
-			logger.Error("streamer failed to create ClientStream", zap.Error(err))
-			return nil, err // TODO(mwitkow): Maybe dial and transport errors should be retriable?
-		}
-		retryingStreamer := &serverStreamingRetryingStream{
-			client:       c,
-			ClientStream: newStreamer,
-			callOpts:     callOpts,
-			ctx:          ctx,
-			streamerCall: func(ctx context.Context) (grpc.ClientStream, error) {
+
+		newCSBuilder, ok := specificCSBuilder[method]
+		if !ok {
+			if callOpts.max == 0 {
 				return streamer(ctx, desc, cc, method, grpcOpts...)
-			},
+			}
+			if desc.ClientStreams {
+				return nil, status.Errorf(codes.Unimplemented, "clientv3/retry_interceptor: cannot retry on ClientStreams without specific interceptor")
+			}
+			newCSBuilder = newGeneralServerCSBuilder
 		}
-		return retryingStreamer, nil
+		return newCSBuilder(c, callOpts).Build(ctx, desc, cc, method, streamer, grpcOpts...)
+	}
+}
+
+type watchCSBuilder struct {
+	client *Client
+}
+
+func newWatchCSBuilder(client *Client, opts *options) clientStreamBuilder {
+	return &watchCSBuilder{
+		client: client,
+	}
+}
+
+type watchReauthStream struct {
+	ctx context.Context
+	grpc.ClientStream
+	client   *Client
+	streamer func() (grpc.ClientStream, error)
+	lastSent interface{}
+	mux      sync.RWMutex
+}
+
+func (b *watchCSBuilder) Build(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+	newStream, err := streamer(ctx, desc, cc, method, opts...)
+	if err != nil {
+		b.client.lg.Error("streamer failed to create ClientStream", zap.Error(err))
+		return nil, err
+	}
+	return &watchReauthStream{
+		ctx:          ctx,
+		ClientStream: newStream,
+		client:       b.client,
+		streamer: func() (grpc.ClientStream, error) {
+			return streamer(ctx, desc, cc, method, opts...)
+		},
+	}, nil
+}
+
+func (s *watchReauthStream) SendMsg(m interface{}) error {
+	s.mux.Lock()
+	s.lastSent = m
+	s.mux.Unlock()
+	return s.ClientStream.SendMsg(m)
+}
+
+func (s *watchReauthStream) RecvMsg(m interface{}) (err error) {
+	if err = s.ClientStream.RecvMsg(m); err != nil {
+		return
+	}
+	// as of 3.4.9 the last sent prior to created response is guaranteed to be corresponding create request
+	if !(s.isLastSentCreateRequest() && s.isAuthInvalidResponse(m)) {
+		return
+	}
+	s.client.lg.Info("refresh auth token and resend watch request")
+	if err = s.client.getToken(s.ctx); err != nil {
```

Comment:
etcd supports multiple authentication methods. If users use other authentication methods, such as certificates. have you tested this scenario(cert has no permission)?
It seems that we need to distinguish token authentication. for example, in unaryClientInterceptor.
```
			if callOpts.retryAuth && rpctypes.Error(lastErr) == rpctypes.ErrInvalidAuthToken {
				gterr := c.getToken(ctx)
				if gterr != nil {
					logger.Warn(
						"retrying of unary invoker failed to fetch new auth token",
						zap.String("target", cc.Target()),
						zap.Error(gterr),
					)
					return gterr // lastErr must be invalid auth token
				}
				continue
			}
```

Reply:
In the case of CN based auth, users won't supply username and password, then `getToken()` will be nop so it will be fine. But caring about different auth mechanisms are good point.

@jschwinger23 could you add e2e test cases for making sure this refreshing mechanism works and it won't break other auth mechanisms? If you are busy adding in other PRs is fine.

Reply:
It seems that it cannot distinguish different auth mechanisms here. I agree with @xkeyideal solution in issue #12157.